src/core/halide2isl.cc

```diff
@@ -82,44 +82,141 @@ isl::aff makeIslAffFromInt(isl::space space, int64_t val) {
   return isl::aff(isl::local_space(space), v);
 }
 
-isl::aff makeIslAffFromExpr(isl::space space, const Expr& e) {
+std::vector<isl::aff> makeIslAffBoundsFromExpr(
+    isl::space space,
+    const Expr& e,
+    bool allowMin,
+    bool allowMax);
+
+namespace {
+/*
+ * Convert Halide binary expression "op" into a list of isl affine functions by
+ * converting its LHS and RHS into lists of affs and concatenating those lists.
+ * This is intended to be used with Min/Max operations in upper/lower bound
+ * computations, respectively.  Essentially, this allows for replacements
+ *   x < min(a,min(b,c)) <=> x < a AND x < b AND x < c
+ *   x > max(a,max(b,c)) <=> x > a AND x > b AND x > c
+ */
+template <typename T>
+inline std::vector<isl::aff>
+concatAffs(isl::space space, T op, bool allowMin, bool allowMax) {
+  std::vector<isl::aff> result;
+
+  for (const auto& aff :
+       makeIslAffBoundsFromExpr(space, op->a, allowMin, allowMax)) {
+    result.push_back(aff);
+  }
+  for (const auto& aff :
+       makeIslAffBoundsFromExpr(space, op->b, allowMin, allowMax)) {
+    result.push_back(aff);
+  }
+
+  return result;
+}
+
+/*
+ * Convert Halide binary expression "op" into an isl affine function by
+ * converting its LHS and RHS into affs and combining them with "combine"
+ * into a single expression.  LHS and RHS are expected to only produce a single
+ * expression.
+ * This is intended for use with operations other than Min/Max that do not
+ * commute nicely in bounds, for example
+ *   x < a + max(b,c)  NOT <=>  x < a + b AND x < a + c for negative values.
+ */
+template <typename T>
+inline isl::aff combineSingleAffs(
+    isl::space space,
+    T op,
+    isl::aff (isl::aff::*combine)(isl::aff) const) {
```

Comment:
quickly eyeballing I can't figure out easily what this ```isl::aff::*combine``` is, care to explain briefly?